### ISSUE TITLE
Re-export crates like `relm4-icons` into the `relm4` crate

### ISSUE THREAD

**Issue by hwittenborn** [Relm4/Relm4#475]
I think it'd be helpful to re-export crates like `relm4-icons` and `relm4-components` into the main `relm4` crate since there all related to each other. Even if they had to be behind a feature flag to enable, I think being able to access everything from one place would make things more ergonimic and cleaner to look at in my code.

**Comment by AaronErhardt** [comment]
Honestly, I'm not sure whether this is possible because both of those crates have `relm4` as dependency. I think this will cause a cyclic dependency problem because `relm4` would need `relm4-icons` which would need `relm4` in the first place.

**Comment by hwittenborn** [comment]
At least with `relm4-icons` it doesn't look like it depends on `relm4`: https://github.com/Relm4/icons/blob/main/Cargo.toml#L22. Is there just something else I'm not seeing?

**Comment by hwittenborn** [comment]
It'd be a bit of a layout change too, but having the `relm4-components` crate just in a module in the main `relm4` create would seem like a clean solution to me. You could put it behind a feature flag like I was saying originally, and then it just wouldn't compile any of that stuff if the user didn't need it.

**Comment by AaronErhardt** [comment]
> At least with relm4-icons it doesn't look like it depends on relm4

That's true, I forgot about that. Yet my first comment is still correct regarding `relm4-components` which cannot be included by `relm4` without causing cyclic dependencies. The only solution for that would be to move `relm4` into a new crate `relm4-core` and then create a new `relm4` crate that only re-exports stuff, but that's not very nice and certainly not worth for just eliminating one line in a `Cargo.toml` configuration.

And the reason why `relm4-icons` is not included is simply that you have to set feature flags for it to enable certain icons, so you always have to include it manually in `Cargo.toml` anyway. There's literally no benefit of re-exporting it from `relm4`.

**Comment by hwittenborn** [comment]
> Yet my first comment is still correct regarding relm4-components which cannot be included by relm4 without causing cyclic dependencies. The only solution for that would be to move relm4 into a new crate relm4-core and then create a new relm4 crate that only re-exports stuff, but that's not very nice and certainly not worth for just eliminating one line in a Cargo.toml configuration.

Would it be a possibility to make `relm4-components` a module in the `relm4` crate and then just get rid of the `relm4-components` crate entirely? I know it's a lot asking for that when there's probably a fair amount of users on the crate right now, but I do think it would be cleaner long-term.

It'd be easy enough to keep `relm4-components` up to date if needed too by just making it contain `use relm4::components::*`.

> And the reason why relm4-icons is not included is simply that you have to set feature flags for it to enable certain icons, so you always have to include it manually in Cargo.toml anyway. There's literally no benefit of re-exporting it from relm4.

The main benefit for me is that it's one less package that I have to keep a version updated for. I know there's tools out there to automate that kind of stuff, but it's just cleaner if I don't have to deal with any of that. It would also help to keep my dependency list lower, since then there's just one package I'm getting everything from.

**Comment by AaronErhardt** [comment]
> Would it be a possibility to make relm4-components a module in the relm4 crate and then just get rid of the relm4-components crate entirely

Yes, that should be possible. We could keep the code behind a feature flag, so it remains optional as it is currently.

> The main benefit for me is that it's one less package that I have to keep a version updated for.

I don't see how that helps. You need to use `relm4-icons = { version = "...", features = ["icon-name"] }` anyway to get any icons to use in your application. Even if you could do something like `relm4 = { ..., features = ["icons"] }` to include `relm4-icons` in `relm4`, you'd still have to include `relm4-icons` manually. If you don't want to update the version, you should be able to set the version to `"*"`.

**Comment by hwittenborn** [comment]
> You need to use `relm4-icons = { version = "...", features = ["icon-name"] }` anyway to get any icons to use in your application. Even if you could do something like `relm4 = { ..., features = ["icons"] }` to include `relm4-icons` in `relm4`, you'd still have to include relm4-icons manually.

How so? If you did something like `pub use relm4-icons` in the `relm4` create, then you don't have to list the `relm4-icons` dependency at all, and `relm4` would just take care of all of that.

> If you don't want to update the version, you should be able to set the version to `"*"`.

At least in my binary crates I like specifying concrete versions for the sake of knowing what versions are used where. I know `Cargo.lock` serves that well, but it's just a personal preference to lock those versions in to `Cargo.toml` as well.

**Comment by hwittenborn** [comment]
I think the stuff that makes me want everything re-exported is because crates like [`libadwaita`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/#reexports) and [`gtk4`](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/#reexports) do it. Not having to specify dependencies like `glib` (and even `gtk4` when using `libadwaita`) really helps to ensure I manage the version of one dependency, and then the rest just gets taken care of.

**Comment by AaronErhardt** [comment]
Have you read the [README](https://github.com/Relm4/icons#2-get-relm4-icons-) of `relm4-icons`? If you just include it without any feature flags, it does nothing. You have to choose the icons you need, add them as feature flags and then they will be compiled into the crate. Just re-exporting does not help at all because you have no icons then. 

I understand that re-exports are great and I'm on the same page with you regarding this generally, but `relm4-icons` does not make sense as a re-export because you must include it anyway to get the icons you need.

**Comment by hwittenborn** [comment]
I had, I was thinking all of the icons could just be enabled if you were re-exporting. I'm seeing how that could be a bad idea for binary size though, and it definitely doesn't provide the same flexibility as using the crate directly.

Looking at my own packages though, the main GTK application I work on is ~80MB, so adding in the 20MB for all the icons (that's what running `du -sh .` on the repository showed anyway) wouldn't be too much of a difference, at least for me.

I'm not feeling as strong as getting that one into the main crate though, that lack of flexibility definitely might not be wanted by a fair amount of people. At the least I think the `component` stuff would still be good to have in the main crate though.

**Comment by hwittenborn** [comment]
Hey @AaronErhardt, could we look back into exporting `relm4-icons` from the main crate, now that the icons crate utilizes a config file instead of features?

**Comment by AaronErhardt** [comment]
It's true that `relm4-icons` is now trivial to include into the core relm4 crate. However, I plan work on a successor to relm4-icons and would like to deprecate it once the replacement is ready. The thing is, while it works great on most setups, the workaround I came up with to reduce the amount of feature flags to satisfy the new demands of the crates.io team is very brittle and is fundamentally broken if you use a shared build directory. I'm planning something like relm4-template, but as a utility that keeps all things up-to-date (flatpak, meson, config.rs etc.) and also builds the bundle for icons (which relm4-icons does currently).

**Comment by hwittenborn** [comment]
That's fair enough, and I'm seeing [the issue for that](https://github.com/Relm4/icons/issues/12) in the `relm4-icons` repository. What specifically are you thinking about for Flatpak/Meson to warrant a new tool? Or if you don't have much fleshed out, that's understandable too.

I'm assuming you've tried pretty much everything to get shared build directories working, but I'm open to looking into that if you think things could still work well with `relm4-icons` as a standalone tool.

**Comment by AaronErhardt** [comment]
> What specifically are you thinking about for Flatpak/Meson to warrant a new tool?

The problem with relm4-template (and gtk-rs-template for that matter) is that you generate your template once and then you have to edit everything by hand. If you want to upgrade the runtime to GNOME 45 for example, you have to edit Cargo.toml, your flatpak manifest(s) and your CI configs. Also, you don't get any improvements added to the template after you generated your initial code. To fix that, I would like to create a tool that manages your boilerplate nicely and it would be easy to add the functionality of `relm4-icons` there as well.

> I'm assuming you've tried pretty much everything to get shared build directories working

I don't know whether there is a solution. I've looked really deep into the rustc/cargo docs, but couldn't find any useful parameters or env variables there. The current approach is to walk up the build directory until you find a `Cargo.toml` file, but that doesn't work for shared build directories of course. AFAIK there is simply no proper way of integrating user configuration into your build process except for feature flags.

> Any thoughts on getting relm4-components integrated into the relm4 crate as well?

If you can make that work, go for it. As said before, this first needs a fix for the cyclic-dependencies, but we could add a `relm4-core` crate for that purpose (I just personally think this is not worth the effort).

**Comment by hwittenborn** [comment]
> I've looked really deep into the rustc/cargo docs, but couldn't find any useful parameters or env variables there.

I'll take your word on that; thanks for the info.

> If you can make that work, go for it. As said before, this first needs a fix for the cyclic dependencies, but we could add a `relm4-core` crate for that purpose (I just personally think this is not worth the effort).

What I was thinking of is moving the `relm4-components` crate into being a module inside of `relm4` entirely, so that it simply wouldn't be separated. For backward compatibility, the `relm4-components` crate could re-export items from `relm4::components::*` then. Are you open to that?
